aten/src/THC/THCTensorMath.cuh

```diff
@@ -40,91 +40,92 @@ inline bool getCatGrid(THCState* state, ptrdiff_t nTensors, dim3& grid) {
-  *
-  * The most important assumption made is that the input tensors are contiguous.
-  */
-
-
-
+ * Kernel used to concatenated grimDim.y tensors into an output tensor. Uses a
+ * grid-stride loop based off of the blockIdx.x, threadIdx.x for each input to
+ * copy each element from each input tensor into the output.
+ *
+ * output: base pointer to the storage associated with the output tensor
+ * inputs: GPU-allocated array of input metadata for each input to concatenate
+ * in the kernel os: the size/stride vectors for the output tensor concatDim:
+ * dimension along which we are concatenating dimStride: the stride of the
+ * output tensor at the concatDim
+ *
+ * The most important assumption made is that the input tensors are contiguous.
+ */
 template <typename T, typename IndexType, int Dims>
 __global__ void CatArrayBatchedCopy(
     T* output,
-    CatArrInputTensor<T, IndexType>* inputs,
-    OutputTensorSizeStride<IndexType, CAT_ARRAY_MAX_INPUT_DIMS> os,
+    CatArrInputTensor<T, IndexType, CAT_ARRAY_MAX_INPUT_DIMS>* inputs,
+    TensorSizeStride<IndexType, CAT_ARRAY_MAX_INPUT_DIMS> os,
     const int concatDim,
-    IndexType dimStride) {
-
-    IndexType tid = blockIdx.x * blockDim.x + threadIdx.x;
-    IndexType nElements = inputs[blockIdx.y].nElements;
-
-    if(tid >= nElements) return;
-    
-    T* data = inputs[blockIdx.y].input;
-    IndexType offset = inputs[blockIdx.y].offset;
-    IndexType dimSize = inputs[blockIdx.y].dimSize;
-    IndexType dataOffset = offset * dimStride;
-
-    IndexType stride = gridDim.x * blockDim.x;
-
-    while( tid < nElements){
-    IndexType elementOffset = CatArrIndexToOffset<IndexType, Dims>::compute(
-    	      os.outputSize, os.outputStride, dimSize, concatDim, tid);
-    output[dataOffset + elementOffset] = data[tid];
-
-    tid += stride;
+    const int pad,
+    T pad_value) {
+  IndexType nElementsOutput = inputs[blockIdx.y].nElementsOutput;
+  T* data = inputs[blockIdx.y].input;
+  IndexType dimOffset = inputs[blockIdx.y].offset;
+  IndexType dataOffset = dimOffset * os.tensorStride[concatDim];
+
+  for (IndexType linearIndex = (IndexType)blockIdx.x * blockDim.x + threadIdx.x;
+       linearIndex < nElementsOutput;
+       linearIndex += (IndexType)gridDim.x * blockDim.x) {
+    if (pad) {
+      IndexType inputOffset = 0;
+      IndexType outputOffset = 0;
+      bool inbound = true;
+      for (int i = Dims - 1; i >= 1; --i) {
+        IndexType inputDimSize = inputs[blockIdx.y].inputParam.tensorSize[i];
+        IndexType curDimSize = i == concatDim ? inputDimSize : os.tensorSize[i];
+        IndexType nextDimIndex = linearIndex / curDimSize;
+        IndexType curDimIndex = linearIndex - curDimSize * nextDimIndex;
+        inbound = inbound && curDimIndex < inputDimSize;
+        inputOffset +=
+            curDimIndex * inputs[blockIdx.y].inputParam.tensorStride[i];
+        outputOffset += curDimIndex * os.tensorStride[i];
+        linearIndex = nextDimIndex;
```

Comment:
you can't assign loop variable within a loop, this leads to infinite looping and hangs.